### ISSUE TITLE
[Bug] T.reduce_sum float16 列归约 dim=0 结果错误（srcRepStride 无法表达 2 字节步进）

### ISSUE THREAD

**Issue by 2Sheldonlaw** [tile-ai/tilelang-ascend#1683]
## Bug Description

`T.reduce_sum` 在 ascendc 后端对 float16 做列归约（dim=0）时，结果完全错误。

根因是 tilelang 对 `float16 + sum` 的 workaround 使用 `WholeReduceSum`，其 `srcRepStride` 参数单位是 **32 字节块**。列归约需要每次 repeat 步进 1 个 float16 = 2 字节，但 `srcRepStride` 的最小单位是 32 字节——无法表达 2 字节步进。`srcRepStride=1`（32 字节）导致每次 repeat 跳过 16 个元素，读的是同一行的连续片段而非同一列的跨行元素。

**只影响 `float16 + sum + dim=0`**，不影响：
- `float16 + max/min + dim=0`（走 `WholeReduceMax/Min`，CANN 原生支持 half，不需要 workaround）
- `float16 + sum + dim=-1`（行归约，`srcRepStride` 计算正确）
- `float32 + sum + dim=0`（走 `ReduceSum<float, RA>`，CANN 原生支持）

**仓库 examples 不触发此问题**——所有 examples 中 `dim=0` 的 reduce_sum 都用 float32。`unsorted_segment_sum` 看似用了 float16，但实际在调用前先将 float16 cast 到 float32（`unsorted_segment_sum.py:185-187`：`if orig_dtype in (torch.float16, torch.bfloat16): data_2d = data_2d.to(torch.float32)`），不触发 Bug #16。

## Environment

- NPU: Ascend910_9362 (A3, dav-c220)
- CPU Architecture: aarch64
- CANN: 9.0.0
- torch: 2.7.1+cpu
- torch-npu: 2.7.1.post4
- Python: 3.11.4
- TileLang-Ascend: `5149acdb40b086d7cd84e30821f5759fa8699a14`
- Installation: build from source

## Code to Reproduce

```bash
source set_env.sh
python repro_reduce_bug16.py
```

<details>
<summary>完整复现代码</summary>

```python
import torch
import tilelang
import tilelang.language as T

tilelang.cache.clear_cache()

M, N = 16, 128


def make_reduce(op_name, dtype, dim):
    @tilelang.jit(out_idx=[0], target="ascendc")
    def kernel():
        @T.prim_func
        def main(B: T.Tensor((N if dim == 0 else M,), dtype)):
            with T.Kernel(1, is_npu=True) as (cid, vid):
                a_ub = T.alloc_ub((M, N), dtype)
                b_ub = T.alloc_ub((N if dim == 0 else M,), dtype)
                T.tile.arith_progression(a_ub, 0.0, 1.0, M * N)
                if op_name == "sum":
                    T.reduce_sum(a_ub, b_ub, dim=dim)
                elif op_name == "max":
                    T.reduce_max(a_ub, b_ub, dim=dim)
                T.copy(b_ub, B)
        return main
    return kernel()


def run_test(op_name, dtype, dim):
    torch_dtype = {"float16": torch.float16, "float32": torch.float32}[dtype]
    print(f"\n--- {op_name} {dtype} dim={dim} ---")
    kernel = make_reduce(op_name, dtype, dim)
    b = kernel()
    torch.npu.synchronize()
    a = torch.arange(M * N, dtype=torch.float32).reshape(M, N)
    if dim == 0:
        ref = a.to(torch_dtype).sum(dim=0) if op_name == "sum" else a.to(torch_dtype).max(dim=0).values
    else:
        ref = a.to(torch_dtype).sum(dim=-1) if op_name == "sum" else a.to(torch_dtype).max(dim=-1).values
    match = torch.equal(b.cpu(), ref.cpu())
    max_diff = (b.cpu().float() - ref.cpu().float()).abs().max().item()
    print(f"  {'PASS' if match else 'FAIL'} (max_diff={max_diff})")
    if not match:
        print(f"  b[:4]: {b.cpu()[:4].tolist()}")
        print(f"  ref[:4]: {ref.cpu()[:4].tolist()}")

print(f"=== Bug #16: float16 sum dim=0 ===")
print(f"M={M}, N={N}, V-pipe src (arith_progression)")

run_test("sum", "float16", 0)    # Bug: FAIL
run_test("max", "float16", 0)    # Control: PASS
run_test("sum", "float32", 0)    # Control: PASS
run_test("sum", "float16", -1)   # Control: PASS (row reduce)
```

</details>

### Reproduction results

```
--- sum float16 dim=0 ---
  FAIL (max_diff=17391.0)
  b[:4]: [120.0, 376.0, 632.0, 888.0]
  ref[:4]: [15360.0, 15376.0, 15392.0, 15408.0]

--- max float16 dim=0 ---
  PASS (max_diff=0.0)

--- sum float16 dim=-1 ---
  PASS
```

float16+sum+dim=0 FAIL，对照组（float16+max+dim=0、float16+sum+dim=-1）PASS。

## Error Message

无编译错误、无运行时错误——**静默失败**。编译通过，运行通过，但结果完全错误。

## Expected Behavior

`T.reduce_sum` 应正确处理 `float16 + dim=0`（列归约）。

## Actual Behavior

codegen 对 `float16 + sum + clear=True` 走专门的 `reduce_sum_half` workaround（因为 CANN `ReduceSum<half, RA>` 的 `static_assert` 拒绝 half）。生成的代码：

```cpp
// dim=0 (FAIL): srcRepStride=1，即 32 字节块步进
tl::ascend::reduce_sum_half<half>(b_ub[0], a_ub[0], 16, 128, 1);
//                                              mask=16, repeat=128, srcRepStride=1
// srcRepStride=1 → 32 字节 = 16 个 float16
// 每次 repeat 跳过 16 个元素（应为 1 个），读同一行而非跨列

// dim=-1 (PASS): srcRepStride=8，即 256 字节步进 = 128 个 float16
tl::ascend::reduce_sum_half<half>(b_ub[0], a_ub[0], 128, 16, 8);
//                                              mask=128, repeat=16, srcRepStride=8
// srcRepStride=8 → 256 字节 = 128 个 float16 = 一整行，正确步进
```

`reduce_sum_half` 实现（`common.h:457-462`）：

```cpp
template <typename T>
CATLASS_DEVICE void
reduce_sum_half(LocalTensor<T> const &dstTensor, LocalTensor<T> const &srcTensor,
                const int32_t mask, const int32_t repeatTime, const int32_t srcRepStride) {
    AscendC::WholeReduceSum<T>(dstTensor, srcTensor, mask, repeatTime, 1, 1, srcRepStride);
    // WholeReduceSum 的 srcRepStride 单位是 32 字节块
}
```

codegen 计算逻辑（`codegen_ascend.cc:2160-2183`）：

```cpp
if (dtype == "half" && clear) {
    if (dim_val == -1) {
        // 行归约：srcRepStride = (n + 15) / 16 = 8  ✓ (128 float16 = 256B = 8 个 32B 块)
    } else if (dim_val == 0) {
        // 列归约：srcRepStride = (m + 15) / 16 = 1  ❌
        // 需要 2 字节步进（1 个 float16），但 srcRepStride=1 表示 32 字节
        // 32 字节 = 16 个 float16，跳过 15 个元素
    }
}
```

## Suggested Solution

对 `half + dim=0` 改用 cast half→float → `ReduceSum<float, RA>` → cast float→half：

```cpp
// 在 codegen 中，对 half + dim=0 不走 reduce_sum_half
// 而是生成：
// 1. Cast<half, float>(src_float, src_half, count)
// 2. ReduceSum<float, RA>(dst_float, src_float, ...)
// 3. Cast<float, half>(dst_half, dst_float, count)
```

或者在 `common.h` 中添加 `reduce_sum_half_col` 函数，封装上述 cast → reduce → cast 流程。

## Additional Context

**Backend:**
- [x] Ascend C backend
- [ ] PTO backend (pto 后端在 dav-c220 上完全不可用，无法测试)
- [ ] AscendNPU IR backend

**影响范围**：仅 `float16 + reduce_sum + dim=0 + clear=True`（ascendc 后端）

**不影响**：
- `float16 + reduce_max/min + dim=0`（走 `WholeReduceMax/Min`，CANN 原生支持 half）
- `float16 + reduce_sum + dim=-1`（行归约，srcRepStride 正确）
- `float32 + reduce_sum + dim=0`（走 `ReduceSum<float, RA>`，CANN 原生支持）

<details>
<summary>源码调用链</summary>

### 为什么需要 workaround

CANN `AscendC::ReduceSum<half, Pattern::RA>` 在所有架构上都不支持 half：

```cpp
// v220 (dav-c220 / 910B) 路径：
// CANN: asc/impl/adv_api/detail/reduce/reduce_sum/reduce_sum_v220_impl.h:287
static_assert(SupportType<T, float>(),
    "failed to check the data type, current api supports data type is float!");

// c310 (dav-3510 / 910C) 路径：
// CANN: asc/impl/adv_api/detail/reduce/reduce_sum/reduce_sum_c310_impl.h:116
static_assert(SupportType<T, int32_t, uint32_t, float, int64_t, uint64_t>(),
    "ReduceSum only support int32_t/uint32_t/float/int64_t/uint64_t data type on current device!");
```

两个架构的 `static_assert` 都不含 half。tilelang 被迫对 `half + sum + clear=True` 走 `WholeReduceSum`（后者支持 half）作为 workaround，但 `WholeReduceSum` 只能读连续数据，适合行归约，不适合列归约。

### Codegen 代码路径

`src/target/codegen_ascend.cc:2160-2183` — `ReduceOpCodegen` 中 `half + clear` 分支：

```cpp
if (dtype == "half" && clear) {
    if (dim_val == -1) {
        srcRepStride = (n_val + 15) / 16;  // 行归约：正确
    } else if (dim_val == 0) {
        srcRepStride = (m_val + 15) / 16;  // 列归约：错误（32B 块无法表达 2B 步进）
    }
    this->stream << "tl::ascend::reduce_sum_half<half>(" << ... << srcRepStride << ");\n";
}
```

### C++ 模板

`src/tl_templates/ascend/common.h:455-462` — `reduce_sum_half`：

```cpp
template <typename T>
CATLASS_DEVICE void
reduce_sum_half(LocalTensor<T> const &dstTensor, LocalTensor<T> const &srcTensor,
                const int32_t mask, const int32_t repeatTime, const int32_t srcRepStride) {
    AscendC::WholeReduceSum<T>(dstTensor, srcTensor, mask, repeatTime, 1, 1, srcRepStride);
}
```

`WholeReduceSum` 的 `srcRepStride` 参数单位是 32 字节块（CANN `ReduceRepeatParams` 文档：`src Stride Unit is 32B`）。

### Examples 规避方式

所有 examples 中 `dim=0` 的 reduce_sum 都用 float32：
- `examples/HISA/block_sparse_mqa_attn_expert_test.py:484` — `T.reduce_sum(s_ub_4x, logits_4x, dim=0, clear=True)`，dtype=float32
- `examples/deepseek_v4/hc_split_sinkhorn.py:120` — `T.reduce_sum(comb_shared, col_sum, dim=0, ...)`，dtype=float32
- `examples/tile_kernels/mhc/head_compute_mix_kernel.py:147` — `T.reduce_sum(sig_ub, reduce_col_ub, dim=0)`，dtype=float32
- `examples/unsorted_segment_sum/unsorted_segment_sum.py:107` — `T.reduce_sum(data_ub, acc_ub, dim=0, clear=True)`，看似 float16 但实际在 line 185-187 先 cast 到 float32：`if orig_dtype in (torch.float16, torch.bfloat16): data_2d = data_2d.to(torch.float32)`

</details>
